onnxruntime/core/optimizer/noop_elimination.cc

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/optimizer/noop_elimination.h"
+
+#include "core/common/logging/logging.h"
+#include "core/graph/graph_viewer.h"
+#include "core/graph/op.h"
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/rewrite_rule.h"
+
+namespace onnxruntime {
+
+/**
+  Eliminate no op node - handling Add op for now
+  Add example: 
+
+      X     0
+       \   /   
+        Add
+         |
+         Y    
+ */
+Status NoopElimination::Apply(Graph& graph, Node& node, RewriteRuleEffect& rule_effect, const logging::Logger&) const {
+  if (graph_utils::RemoveNode(graph, node)) {
+    rule_effect = RewriteRuleEffect::kRemovedCurrentNode;
+  }
+
+  return Status::OK();
+}
+
+bool NoopElimination::SatisfyCondition(const Graph& graph, const Node& node, const logging::Logger& logger) const {
+
+  bool input0_is_initializer = graph_utils::IsConstantInitializer(graph, node.InputDefs()[0]->Name());
+  bool input1_is_initializer = graph_utils::IsConstantInitializer(graph, node.InputDefs()[1]->Name());
+
+  // reject if both or neither inpuits are initializers for now
+  if (input0_is_initializer == input1_is_initializer) {
+    return false;
+  }
+
+  const auto* initializer = graph_utils::GetConstantInitializer(graph, node.InputDefs()[input0_is_initializer ? 0 : 1]->Name());
+
+  int32_t data_type = initializer->data_type();
```

Comment:
Can you verify what happens if the constant node has shape [1][1][1] and data=0 and the other node has shape [1][1]? With broadcasting rules, the final shape should be [1][1][1], but that wouldn't happen here. Maybe add a check that the initializer rank is < non_initializer rank?

Reply:
in this case, it might be optimal perf-wise to replace the Add with an Unsqueeze. This might be outside of the scope of this PR though.